src/circuit/ecdsa.rs

```diff
@@ -0,0 +1,339 @@
+/// E is the emulated curve, C is the native curve
+struct EcdsaChip<E: CurveAffine, C: CurveAffine> {
+    pub config: EcdsaConfig,
+    pub rns_base_field: Rns<E::Base, C::ScalarExt>,
+    pub rns_scalar_field: Rns<E::Scalar, C::ScalarExt>,
+}
+
+impl<E: CurveAffine, C: CurveAffine> EcdsaChip<E, C> {
+    pub fn new(config: EcdsaConfig, rns_base_field: Rns<E::Base, C::ScalarExt>, rns_scalar_field: Rns<E::ScalarExt, C::ScalarExt>) -> Result<Self, Error> {
+        Ok(
+            Self {
+                config,
+                rns_base_field,
+                rns_scalar_field,
+            }
+        )
+    }
+
+    pub fn configure(_: &mut ConstraintSystem<C::ScalarExt>, ecc_chip_config: &EccConfig) -> EcdsaConfig {
+        EcdsaConfig {
+            ecc_chip_config: ecc_chip_config.clone(),
+        }
+    }
+
+    pub fn ecc_chip(&self) -> Result<GeneralEccChip<E, C::ScalarExt>, Error> {
+        let ecc_chip_config = self.config.ecc_chip_config();
+        GeneralEccChip::new(ecc_chip_config, self.rns_base_field.clone(), self.rns_scalar_field.clone())
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct EcdsaSig<'a, W: FieldExt, N: FieldExt> {
+    pub r: Integer<'a, W, N>,
+    pub s: Integer<'a, W, N>,
+}
+
+pub struct AssignedEcdsaSig<N: FieldExt> {
+    pub r: AssignedInteger<N>,
+    pub s: AssignedInteger<N>,
+}
+
+pub struct AssignedPublicKey<N: FieldExt> {
+    pub point: AssignedPoint<N>,
+}
+
+impl<E: CurveAffine, C: CurveAffine> EcdsaChip<E, C> {
+
+    // https://en.wikipedia.org/wiki/Elliptic_Curve_Digital_Signature_Algorithm
+    fn verify(
+        &self,
+        region: &mut Region<'_, C::ScalarExt>,
+        sig: &AssignedEcdsaSig<C::ScalarExt>,
+        pk: &AssignedPublicKey<C::ScalarExt>,
+        msg_hash: &AssignedInteger<C::ScalarExt>,
+        offset: &mut usize,
+    ) -> Result<(), Error> {
+        let ecc_chip = self.ecc_chip()?;
+        let scalar_chip = ecc_chip.scalar_field_chip();
+        let base_chip = ecc_chip.base_field_chip();
+
+        // 1. check 0 < r, s < n
+
+        // // since `assert_not_zero` already includes a in-field check, we can just call `assert_not_zero`
+        scalar_chip.assert_not_zero(region, &sig.r, offset)?;
+        scalar_chip.assert_not_zero(region, &sig.s, offset)?;
+
+        // 2. w = s^(-1) (mod n)
+        let (s_inv, _) = scalar_chip.invert(region, &sig.s, offset)?;
+
+        // 3. u1 = m' * w (mod n)
+        let u1 = scalar_chip.mul(region, &msg_hash, &s_inv, offset)?;
+
+        // 4. u2 = r * w (mod n)
+        let u2 = scalar_chip.mul(region, &sig.r, &s_inv, offset)?;
+
+        // 5. compute Q = u1*G + u2*pk
+        let e_gen = ecc_chip.assign_point(region, Some(E::generator()), offset)?;
+        let g1 = ecc_chip.mul_var(region, e_gen.clone(), u1, offset)?;
+        let g2 = ecc_chip.mul_var(region, pk.point.clone(), u2, offset)?;
+        let q = ecc_chip.add(region, &g1, &g2, offset)?;
+
+        // 6. check if Q.x == r (mod n)
+        let q_x = q.get_x();
```

Comment:
In order to make x coordinate scalar field element by `x % n` we can just use reduction in scalar field chip. Then or `assert_strict_equal` must work against `sig.r`

Reply:
thanks, now it's done